### PR TITLE
WIP: Added F# templates to 0.10 release

### PR DESCRIPTION
**Pull request by grokys**
Adds F# templates to https://github.com/AvaloniaUI/avalonia-dotnet-templates/pull/45

### CODE REVIEW

**Comment by grokys** [comment]
@jaggerjo @FoggyFinder I've made a start on porting the F# templates from the VS extension here, however I've hit a snag in that I'm getting this error and I don't know enough F# to know how to fix it best:

```
D:\projects\AvaloniaUI\avalonia-dotnet-templates\tests\output\F#\AvaloniaMvvm\Program.fs(12,9): error FS0501: The member or object constructor 'LogToTrace' takes 2 argument(s) but is here given 0. The required signature is '(extension) AppBuilder.LogToTrace<'T when 'T : (new : unit -> 'T) and 'T :> Controls.AppBuilderBase<'T>>(?level: Logging.LogEventLevel, [<ParamArray>] areas: string []) : 'T'. [D:\projects\AvaloniaUI\avalonia-dotnet-templates\tests\output\F#\AvaloniaMvvm\AvaloniaMvvm.fsproj]
```

What should I do here?

Also I've modified the VS templates to give stuff proper namespaces (they were named just `Views` etc in the VS extension, without an application-level namespace). Not sure what the standard way to do this is in F# but @JaggerJo's templates had this so I assume it's just user preference? If so better to keep them consistent with the C# templates.

If you could also check the templates in general that would be good. There's a test script over at https://github.com/AvaloniaUI/avalonia-dotnet-templates/blob/d30326365fb51719bdd397073e239ffdec6c2830/tests/build-test.ps1 which tests that the templates compile etc if that would be useful.

**Comment by JaggerJo** [comment]
Just checked and the issue is that the `LogToTrace` method takes a non optional param array as the 2nd parameter. F# is a bit strange when it comes to parameter arrays.. 

```fsharp
.LogToTrace(areas = Array.empty) 
or 
.LogToTrace(areas = [| |]) 
```

resolves the issue.

File: templates/fsharp/app-mvvm/App.axaml.fs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+type App() as self =
+    inherit Application()
+
+    do AvaloniaXamlLoader.Load self
+
+    override x.OnFrameworkInitializationCompleted() =
+        match x.ApplicationLifetime with
+        | :? IClassicDesktopStyleApplicationLifetime as desktop ->
+             desktop.MainWindow <- new MainWindow(DataContext=MainWindowViewModel())
+        | _ -> ()
```
Comment (JaggerJo) [inline templates/fsharp/app-mvvm/App.axaml.fs:18]:
```suggestion
    override this.OnFrameworkInitializationCompleted() =
        match this.ApplicationLifetime with
        | :? IClassicDesktopStyleApplicationLifetime as desktop ->
             desktop.MainWindow <- new MainWindow(DataContext = MainWindowViewModel())
        | _ -> ()
```

```suggestion
    override self.OnFrameworkInitializationCompleted() =
        match self.ApplicationLifetime with
        | :? IClassicDesktopStyleApplicationLifetime as desktop ->
             desktop.MainWindow <- new MainWindow(DataContext = MainWindowViewModel())
        | _ -> ()
```

File: templates/fsharp/app-mvvm/Program.fs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+open AvaloniaAppTemplate
+
+module Program =
+
+    [<CompiledName "BuildAvaloniaApp">] 
+    let buildAvaloniaApp () = 
+        AppBuilder.Configure<App>()
+                .UsePlatformDetect()
+                .LogToTrace()
+                .UseReactiveUI()
```
Comment (JaggerJo) [inline templates/fsharp/app-mvvm/Program.fs:15]:
```suggestion
        AppBuilder
            .Configure<App>()
            .UsePlatformDetect()
            .LogToTrace(areas = Array.empty)
            .UseReactiveUI()
```

File: templates/fsharp/app/App.axaml.fs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    inherit Application()
+
+    do AvaloniaXamlLoader.Load self
+
+    override x.OnFrameworkInitializationCompleted() =
+        match x.ApplicationLifetime with
+        | :? IClassicDesktopStyleApplicationLifetime as desktop ->
+             desktop.MainWindow <- new MainWindow()
+        | _ -> ()
+
```
Comment (JaggerJo) [inline templates/fsharp/app/App.axaml.fs:17]:
```suggestion
    override this.OnFrameworkInitializationCompleted() =
        match this.ApplicationLifetime with
        | :? IClassicDesktopStyleApplicationLifetime as desktop ->
             desktop.MainWindow <- new MainWindow()
        | _ -> ()

```

```suggestion
    override self.OnFrameworkInitializationCompleted() =
        match self.ApplicationLifetime with
        | :? IClassicDesktopStyleApplicationLifetime as desktop ->
             desktop.MainWindow <- new MainWindow()
        | _ -> ()

```

File: templates/fsharp/app/Program.fs [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+open Avalonia
+open AvaloniaAppTemplate
+
+module Program =
+
+    [<CompiledName "BuildAvaloniaApp">] 
+    let buildAvaloniaApp () = 
+        AppBuilder.Configure<App>()
+                .UsePlatformDetect()
+                .LogToTrace()
```
Comment (JaggerJo) [inline templates/fsharp/app/Program.fs:13]:
```suggestion
        AppBuilder
            .Configure<App>()
            .UsePlatformDetect()
            .LogToTrace(areas = Array.empty)
```

File: templates/fsharp/app/Program.fs [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+    [<STAThread>][<EntryPoint>]
```
Comment (JaggerJo) [inline templates/fsharp/app/Program.fs:15]:
```suggestion
    [<STAThread>]
    [<EntryPoint>]
```
or
```suggestion
    [<EntryPoint; STAThread>]
```

**Review by JaggerJo — COMMENTED** [review]
🚀

**Comment by FoggyFinder** [comment]
```fs
or
.LogToTrace(areas = [| |]) 
```

another option would be

```fs
.LogToTrace(Logging.LogEventLevel.Warning)
```

**Review by FoggyFinder — COMMENTED** [review]
Does `.axaml` work fine with F#? I wasn't able to make it work on `0.9.x` so I stick to `.xaml`

File: templates/fsharp/app-mvvm/App.axaml.fs (hunk 1 above)
Comment (FoggyFinder) [reply templates/fsharp/app-mvvm/App.axaml.fs:18]: it's against F# style guide:

https://docs.microsoft.com/en-us/dotnet/fsharp/style-guide/formatting#formatting-white-space-in-expressions

Comment (grokys) [reply templates/fsharp/app-mvvm/App.axaml.fs:18]: Ok, I won't apply this one if it's against the style guide.

**Pull request closed without merging**